### ISSUE TITLE
Options in Customizer

### ISSUE THREAD

**Issue by danyj** [ThemeFuse/Unyson#410]
Have you planed to include theme options in customizer?

**Comment by danyj** [comment]
Nothing on this?

**Comment by moldcraft** [comment]
We didn't planed that. But we need to investigate, maybe we will do.

**Comment by albertso** [comment]
I did something similar once. I'll try to recover it and adapt it to your framework if I have time.

As a brief guideline, you should hook into the "[customize_register](https://codex.wordpress.org/Class_Reference/WP_Customize_Manager)" action, loop through the fields registered in Unyson, and then register the necessary settings and controls for each field with $wp_customize->add_control() and $wp_customize->add_setting().

Each type of field requires different setting and controls, so switch on each field type and act differently on each case that requires it. Usually it's one regular setting and one control per field, but more advanced field types (like image selection which uses hidden fields) usually need extra controls/settings.

Also, note the differece in attributes used to register controls, and to register settings. Prepare the attributes accordingly before registering them, and be careful not to mix them. 

You should also think about how options in the customizer are organized in sections, so maybe add "section" as an extra option container type, to box, tab and group, or else add the orphaned options in a "general" section, or simply interpret all container types as a "section".

At least this is how I would approach it. I hope it helps.

**Comment by danyj** [comment]
I think that we should get this one going, slowly but surely themes with customizer are taking over and people are kinda getting tired of the actual theme options without previews.
@albertso if you could show us what you did it would be great. 
yes I was just looking at it and a loop seems like it would work, thing is that we have some really special options in Unyson that need a  port to customizer but looking at customizer instructions I hope that this one would work


	$wp_customize->add_control(
		new FW_Option_Type_Multi_Picker(
			$wp_customize,
			'some-setting',
			array(
				'label' => 'Some Setting',
				'section' => 'example_section_one',
				'settings' => 'some-setting',
			)
		)
	);

derived from here http://themefoundation.com/wordpress-theme-customizer/

these guys have a pretty nice set to 
https://github.com/aristath/kirki and getting ready to make it Redux compatible
and Redux options already work in customizer.

**Comment by danyj** [comment]
If you care to look I am testing the waters now
http://wordpress.stackexchange.com/questions/184671/how-to-tell-theme-customizer-to-save-setting-in-my-option-table

**Comment by danyj** [comment]
Can you believe this 
http://wptavern.com/wordpress-org-now-requires-theme-authors-to-use-the-customizer-to-build-theme-options

**Comment by duongancol** [comment]
In my opinion, it is good for user experience, all users will know how to custom theme using customizer.
Moreover, it's core of wordpress, so it's safe and very security, today you can see a lot of framework option like redux, option tree have problem with XSS Vulnerability, read more here: 
http://wptavern.com/xss-vulnerability-affects-more-than-a-dozen-popular-wordpress-plugins
The customizer is better in each version of wordpress but for me now, I prefer theme option, especial Unyson theme option. Because it's very flexible and strong enough.

**Comment by danyj** [comment]
not saying is bad , but mandatory is bit over top and unyson not having options inside customizer  yet ( I know you working on it ) is kinda show stopper. hopefully TF will not enforce this.

**Comment by duongancol** [comment]
@danyj : TF will not, all top of theme on TF are using their own option framework ;)

**Comment by danyj** [comment]
Who ever is on this task , you want to look at this http://kirki.org/, 
they made sub sub  menus for customizer section which if converted right can transform our option tabs and sub tabs to it . This is what I mean http://screencast.com/t/J0DkLbU5aMW

**Comment by duongancol** [comment]
I used to, this is very useful customizer and easy to use.

**Comment by danyj** [comment]
When  you get a chance , if  you have started on this, please advise if you are planing to port our options to customizer as they are now in theme options , or is there going to be a different array structure and if all options will be able to be ported. I am most concerned about the addable popups, image pickers, multi pickers and other specialty options.  Reason why I am asking is to get prepared and structure the options different way. Or to  look for another solution for customizer. Thank you!

**Comment by danyj** [comment]
@moldcraft please guys , just a hint so that I know what to do, wait or go with custom solution.

**Comment by moldcraft** [comment]
I will take a look at this next week.
Because of the TF requirement, we will try everything to make options work in customizer. I think we will make `framework-customizations/theme/options/customizer.php`

**Comment by danyj** [comment]
thnx for reply , I will wait few more days than. If you cant do this than please do let us know.

**Comment by moldcraft** [comment]
*I wasn't at office the last 8 weeks, that's why some issues were fixed with big delay. In a feew days I will come back to full-time.*

**Comment by duongancol** [comment]
I don't like the theme option in customizer, it should be separated. You should consider use one of theme in your theme before developing. 
Firstly, if you support option in both theme option and customizer, user will not know where to go to setup. It's very confuse. 
Secondly, I think it will limit functions of Theme Option.
Lastly, WP core now can support all most options like Theme option do. I prefer if Unyson support some default fields so that user can intergrate easily like Theme Option.

**Comment by danyj** [comment]
@duongancol  how is it confusion to work with visual editor? that is what everyone is after now. 
only reason x bypassed everyone in weekly sales in last 2 weeks by a looong mile is the customizer. 

and if unyson adds options they should not be base only , first thing everyone will try to switch is typography, font size , backgrounds ,  and if you have set those up in your admin inside a popup , addable popup than what?  if you used multi-picker same thing. so I think they should try their best to port all if possible.  this way any dev can decide , use both theme option or customizer or use them individually. 

what I did 2 months ago is topmenu visual editor in theme options that works perfect , than when I was done with options I  thought that it would be nice to have a complete site visual layout in backend so that while admin works he/she can see the change , you can guess that this took me on the road to a slow admin ,   but if I had a customizer , I would not have to do any of it and would just need to setup the options. and I doubt that I am the only one who wants to see the theme changes before they are saved.

**Comment by danyj** [comment]
oh and I and talking about options functionality ,  not port my complete theme options backend to customizer. I am talking about options to be functional in customizer as they are in all areas of unyson

**Comment by duongancol** [comment]
got ti @danyj  and I agree with that.

**Comment by moldcraft** [comment]
**Options works in customizer**.

Currently all the changes I made are on the [customizer](https://github.com/ThemeFuse/Unyson/tree/customizer) branch.  If you are using git `git pull origin && git checkout customizer`, else click on the previous link then click on **Download ZIP** in bottom right side of the page.

----

**What's in the box**:

* Options for customizer are loaded from

    ```
    {theme}/framework-customizations/theme/options/customizer.php
    ```

* Options array structure is the same as settings, post, taxonomy options. Nothing new added, **yet**.
* **Panels** are created from **Tabs**

    ```php
    'panel_id' => array(
        'type' => 'tab',
        'title' => 'Panel Title',
        'options' => array(
            // This MUST contain 'box' containers
        )
    )
    ```

    *Note: Empty panels are hidden. To test/view a panel, you must place in it sections/boxes and controls/options.*

* **Sections** are created from **Boxes**

    ```php
    'section_id' => array(
        'type' => 'box',
        'title' => 'Section Title',
        'options' => array(
            // This MUST contain options (not containers)
        )
    )
    ```

    *Note: Empty sections are hidden. To test/view a section, you must place in it controls/options.*

* Everything is saved in [theme mods](https://codex.wordpress.org/Function_Reference/get_theme_mod) in `fw_options` key.

    *Nothing custom, the default/native save by WordPress.*

* To get db value, use `$value = fw_get_db_customizer_option('option_id')` instead of

    ```php
    $all_options = get_theme_mod( 'fw_options', array() )

    if (isset($all_options['option_id'])) {
        $value = $all_options['option_id'];
    } else {
        // extract default value from options array
        $all_default_values = fw_get_options_values_from_input(
            fw()->theme->get_customizer_options(),
            array()
        );
        if (isset($all_default_values['option_id'])) {
            $value = $all_default_values['option_id'];
        } else {
            $value = null;
        }
    }

    // use $value ...
    ```

* `group` containers are not supported
* If your options array will contain something wrong, you will get warnings.

----

**What I will do next**:

* Improve options display (html/css fixes)
* Update preview on keyup, addable-option add/remove, etc. *(currently it works only on any input change)*
* A way to hook into an option and make it work with real-time preview instead of full preview refresh.

*I will let you know when I will finish some of the above tasks*

----

**What I am waiting from you**: 

* Test (play with it)
* Tell me if I missed something

**Comment by danyj** [comment]
just got this , will test asap and report back . in mean time question ; why are you saving it in theme mods , those customizer options should be connected to unyson theme options , example if I  am using 

   fw_get_db_settings_option('site_width');

to set site width I should not be changing that now to be 

  fw_get_db_customizer_option('site_width');

I  connected kirki options with unyson  in order to preview unyson option but on save the fw options table is being updated 
https://github.com/reduxframework/kirki/issues/203#issuecomment-95927933

**Comment by danyj** [comment]
or maybe I missed something in explanation , disregard this if I was wrong ,

**Comment by danyj** [comment]
like advised above ,  this option is in theme admin 
http://prntscr.com/72qart

	'site_width' => array(
		'type' => 'short-text',
		'value' => '1400px',
		'label' => __('Site max width', 'fw'),
		'desc' => __('Define website max width', 'fw'),
		'help' => __('You can use pixels ( px ) or percentage ( % ). If no operator is added px is default.', 'fw')
	)



now my customizer is this 



	<?php
	if (!defined('FW'))
		die('Forbidden');
	$options = array(
		'panel_id' => array(
			'type' => 'tab',
			'title' => 'Site',
			'options' => array(
				// This MUST contain 'box' containers
				'section_id' => array(
					'type' => 'box',
					'title' => 'Layout',
					'options' => array(
						// This MUST contain options (not containers)
						'site_width' => array(
							'type' => 'short-text',
							'value' => '1400px',
							'label' => __('Site max width', 'fw'),
							'desc' => __('Define website max width', 'fw'),
							'help' => __('You can use pixels ( px ) or percentage ( % ). If no operator is added px is default.', 'fw')
						)
					)
				)
			)
		)
		
	);


is visible in customizer   http://prntscr.com/72qb3c

does not see the value , and is not updating the same option id in 

    fw_theme_settings_options:theme_name

**Comment by danyj** [comment]
and here is that same option done in kirki with the connection I referenced 
http://screencast.com/t/VBEQlLy8FIF2
and is updating unyson admin on save

**Comment by danyj** [comment]
just wanted to say thank you much again for this. my custom options work on it as well! excellent!

**Comment by danyj** [comment]
hint , modals work ,  just need big  -z-index

**Comment by thednp** [comment]
Mhm, when do we have a possible release?

**Comment by duongancol** [comment]
Tested and there are some problems such as:
1. modal of Upload, Addable Popup didn't display well
2. Date range, date picker didn't show calendar
3, Multi- select did not work

Couldn't think it can be done very quickly, you're amazing

**Comment by danyj** [comment]
@moldcraft ,  please give us such hook 

    add_action('fw_admin_enqueue_scripts:customizer', '_thz_admin_scripts');

**Comment by thednp** [comment]
I would very much like to enable/disable sections in the theme panel for the customizer.

**Comment by danyj** [comment]
if you are talking about default wp ones 


	global  $wp_customize;
	
	if (isset($wp_customize)) {
			
		function _customizer_remove_sections( $wp_customize ) {
	
			$wp_customize->remove_section( 'colors' );
			$wp_customize->remove_section( 'background_image' );
			
	
		}
		add_action( 'customize_register' , '_customizer_remove_sections' );
	}

**Comment by danyj** [comment]
add this also

    $wp_customize->remove_section( 'title_tagline' );

**Comment by danyj** [comment]
@thednp  updated above

**Comment by thednp** [comment]
THanks @danyj

**Comment by thednp** [comment]
@danyj I am talking about both wp and Unyson theme powered options.

**Comment by danyj** [comment]
@thednp  the unyson customizer sees only sections you add to customizer.php  so I am not sure what is there to disable/enable

**Comment by thednp** [comment]
@danyj I have some sections of advanced styling (set border radius for each type of component, or box shadow) or theme information "options". I don't want those sections in the customizer.

**Comment by danyj** [comment]
bud , please try the customizer first than  it will make sense what I am saying 
https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-100037709

the customizer is not automatically adding all your theme options to customizer view 

it shows only the ones you add to customizer.php

**Comment by moldcraft** [comment]
> @danyj https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-100033694
> why are you saving it in theme mods , those customizer options should be connected to unyson theme options

To not make custom hacks and to use default wp behavior. I don't make the save, wordpress does that.

Maybe there will be (new) themes with some options in Customizer, and some options (completely different) in Theme Settings.

For those who want to sync or fallback for already existing themes (when they will migrate from Theme Settings to Customizer) I will do a script that can be placed in theme `hooks.php` that will do the sync (something like in the link you gave https://github.com/reduxframework/kirki/issues/203#issuecomment-95927933 )

> @danyj https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-100045257
> hint , modals work , just need big -z-index

> @duongancol https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-100088524
> 1. modal of Upload, Addable Popup didn't display well
> 2. Date range, date picker didn't show calendar

I fixed z-index for modals and date-picker.

@duongancol I will take a look at **3.**

> @danyj https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-100113396
> please give us such hook
>
>     add_action('fw_admin_enqueue_scripts:customizer', '_thz_admin_scripts');

Done

**Comment by danyj** [comment]
great news , thank you!,  please consider this while making the connect class , 
the class I have from kirki is raw and updates complete set of options via clone all options , on save copy clone to fw options, 
now in my theme admin I might have  5 options , but will use only 3 of them in customizer, 
those 3 are only the ones that should be updated. hm , maybe the clone already does that in this way , I did not digg to much in it  so far, just tested with one or 2 options. 

but just look closely please and consider adding option for custom clone db name like kirki. 
again , for the second time , why dont you have donation button ? if you do , where is it?

**Comment by moldcraft** [comment]
> please consider this while making the connect class , 
> the class I have from kirki is raw and updates complete set of options via clone all options

I was thinking to do something like this

```php
// on customizer options save

foreach (
    array_intersect_key(
        $customizer_options_values,
        fw_extract_only_options(fw()->theme->get_settings_options())
    ) 
    as $option_id => $option_value
) {
    fw_set_db_settings_option($option_id, $option_value);
}
```

> again , for the second time , why dont you have donation button ? if you do , where is it?

The best donation would be to tell the world about this framework: write in your blog about it, recommend on forums, etc.

**Comment by duongancol** [comment]
:+1:

**Comment by danyj** [comment]
@moldcraft  , excelent , that is the idea , update only the ones changed , not the whole group. 

as for pushing Unyson , our 100k+ members are waiting for this , so when it comes out it will be a big push. By looking at current unyson community I see few devs working on big things so I am sure that once we all release what we are working on , you my friend will have your hands full ! I just hope that your team will be able to manage it.

**Comment by danyj** [comment]
and when I say manage it  I mean manage the framework itself , we will take care of our docs and customers of course.

**Comment by moldcraft** [comment]
> https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-99988429
> * Update preview on keyup, addable-option add/remove, etc. (currently it works only on any input change)

Done

> https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-99988429
> * A way to hook into an option and make it work with real-time preview instead of full preview refresh.

I have nothing to do here. The default [WordPress solution](https://codex.wordpress.org/Theme_Customization_API#Part_3:_Configure_Live_Preview_.28Optional.29) works.

An example:

*  Change setting transport and enqueue the javascript

    ```php
    // file: {theme}/inc/hooks.php

    if (defined('FW')):
    	/**
    	 * @param WP_Customize_Manager $wp_customize
    	 * @internal
    	 */
    	function _action_customizer_live_fw_options($wp_customize) {
    		if ($wp_customize->get_setting('fw_options[OPTION_ID]')) {
    			$wp_customize->get_setting('fw_options[OPTION_ID]')->transport = 'postMessage';
    
    			add_action( 'customize_preview_init', '_action_customizer_live_fw_options_preview' );
    		}
    	}
    	add_action('customize_register', '_action_customizer_live_fw_options');
    
    	/**
    	 * @internal
    	 */
    	function _action_customizer_live_fw_options_preview() {
    		wp_enqueue_script(
    			'mytheme-customizer',
    			get_template_directory_uri() .'/assets/js/theme-customizer.js',
    			array( 'jquery','customize-preview' ),
    			fw()->theme->manifest->get_version(),
    			true
    		);
    	}
    endif;
    ```

* Handle the change in javascript

    ```javascript
    // file: {theme}/assets/js/theme-customizer.js

    ( function( $ ) {
    	wp.customize( 'fw_options[OPTION_ID]', function( value ) {
    		value.bind( function( newval ) {
    			/**
    			 * An array of collected html inputs
    			 * [{'name':'input[name]','value':'input value'}]
    			 * or
    			 * [{'name':'input[name]','value':'input value'},{'name':'input[name]','value':'input value'},...]
    			 */
    			newval = JSON.parse(newval);
    
    			$( 'h1' ).text( newval[0].value );
    		} );
    	} );
    } )( jQuery );
    ```

**Comment by duongancol** [comment]
Tested and all it worked well. :100:

**Comment by danyj** [comment]
Thnx for the improvements!

In reference to  

    'transport' => 'postMessage'



kirki makes this much more easier without any need to do JS 




		'type'        => 'text',
		'setting'     => 'site_width',
		'label'       => __( 'Site width', 'kirki' ),
		'description' => __( 'Desc', 'kirki' ),
		'help'        => __( 'Help', 'kirki' ),
		'section'     => 'site_layout_section',
		'default'     => thz_get_option('site_width'),
		'priority'    => 10,
		'output'      => array(
			array(
				'element'  => '.thzsitew', // class/tag/id  of the element to affect
				'property' => 'max-width',// what css property to affect
				//'units'	   => ' !important' // addtional css units
				
			)
		),
		
		'transport' => 'postMessage', //  transport type 
		'js_vars'   => array(
			array(
				'element'  => '.thzsitew',  // class/tag/id  of the element to affect
				'function' => 'css', // type of style
				'property' => 'max-width', // what css property to affect
			),
		)


this way we dont have to write additional js and if you are adding another option you dont have to think of it. 

But I do understand that you dont want to change default option types to add transport or other attributes

**Comment by moldcraft** [comment]
I merged the customizer branch into master.
The only thing that remained is to review options design (I will do that with our designer tomorrow).

Is there anything else that needs to be done here? (In case I missed something)

**Comment by danyj** [comment]
@moldcraft is the connection for default DB values already in?

**Comment by moldcraft** [comment]
*That's what I forgot.*
I will do the script that sync common customizer an theme settings, we also need it for our existing theme.

> connection for default DB values

or you are talking about something else?

**Comment by danyj** [comment]
@moldcraft   that is what I meant.

**Comment by danyj** [comment]
this one here https://github.com/reduxframework/kirki/issues/203#issuecomment-95927933

**Comment by danyj** [comment]
do me a favor if you could , 
redo  page-builder style for customizer , let it be same as default design vise just smaller boxes, 
current one works but  all columns and rows and 100% width  and see if you can trigger the change when box is added.

**Comment by moldcraft** [comment]
A builder in that little space in customizer? I think this is not practical.

**Comment by moldcraft** [comment]
100% is because of the `.fw-force-xs` class, in which I wrapped all elements to make them look like the mobile version.

**Comment by danyj** [comment]
I know , never mind I am using builder for something else so I will adjust mine but the trigger and functionality would be grateful.  it works currently but it is not saving and it does not trigger the change

**Comment by moldcraft** [comment]
> it does not trigger the change

Try [this](https://github.com/ThemeFuse/Unyson-Builder-Extension/commit/4177237ba0b614dd8899fcd7c4eb86a566e08ad2).

**Comment by danyj** [comment]
no no , save and trigger works perfect in theme options,  it does not  save  when inside customizer, 
I also tried your page-builder option and is doing the same

**Comment by danyj** [comment]
@moldcraft  oh , you just changed it , will test

**Comment by moldcraft** [comment]
I understood that it happens in customizer. Have you tried [this fix](https://github.com/ThemeFuse/Unyson-Builder-Extension/commit/4177237ba0b614dd8899fcd7c4eb86a566e08ad2) ?

**Comment by danyj** [comment]
ok , triggers now , but still not saving

**Comment by moldcraft** [comment]
Just info for you to understand how the customizer option value works.

[ ![s](https://static.md/82310653016ca0094cfea8e6f69fd078.png) ](https://static.md/82310653016ca0094cfea8e6f69fd078.png)

----

> but still not saving

I will inspect that.

**Comment by moldcraft** [comment]
updated image

**Comment by danyj** [comment]
thnx for the heads up , I'm  ok with those , just now I busy finishing the admin and testing customizer only when you make changes. once done with admin il dig trough all of these better and test more. sure let me know if there are specific I should test for you while you work.

**Comment by moldcraft** [comment]
```php
// file: {theme}/inc/hooks.php


/**
 * Sync common Theme Settings and Customizer options db values
 * @internal
 */
class _FW_Theme_Sync_Customizer_Options {
	public static function init() {
		add_action('customize_save_after', array(__CLASS__, '_action_after_customizer_save'));

		add_action('fw_settings_form_saved', array(__CLASS__, '_action_after_settings_save'));
		add_action('fw_settings_form_reset', array(__CLASS__, '_action_after_settings_save'));
	}

	/**
	 * If a customizer option also exists in settings options, copy its value to settings option value
	 */
	public static function _action_after_customizer_save()
	{
		$settings_options = fw_extract_only_options(fw()->theme->get_settings_options());

		foreach (
			array_intersect_key(
				fw_extract_only_options(fw()->theme->get_customizer_options()),
				$settings_options
			)
			as $option_id => $option
		) {
			if ($option['type'] === $settings_options[$option_id]['type']) {
				fw_set_db_settings_option(
					$option_id, fw_get_db_customizer_option($option_id)
				);
			}
		}
	}

	/**
	 * If a settings option also exists in customizer options, copy its value to customizer option value
	 */
	public static function _action_after_settings_save()
	{
		$customizer_options = fw_extract_only_options(fw()->theme->get_customizer_options());

		foreach (
			array_intersect_key(
				fw_extract_only_options(fw()->theme->get_settings_options()),
				$customizer_options
			)
			as $option_id => $option
		) {
			if ($option['type'] === $customizer_options[$option_id]['type']) {
				fw_set_db_customizer_option(
					$option_id, fw_get_db_settings_option($option_id)
				);
			}
		}
	}
}
_FW_Theme_Sync_Customizer_Options::init();
```

**Comment by danyj** [comment]
ok but not even simple input works for me ,  

how  do you setup option in custoimizer 



this 

					'site_width' => array(
						'type' => 'short-text',
						'value' => '1400px',
						'label' => __('Site max width', 'fw'),
						'desc' => __('Define website max width', 'fw'),
						'help' => __('You can use pixels ( px ) or percentage ( % ). If no operator is added px is default.', 'fw')
					),


is same option from theme backend and to me it does not return value

http://prntscr.com/74pfpk

**Comment by danyj** [comment]
this is from admin , same option 
http://prntscr.com/74pfzn

**Comment by moldcraft** [comment]
The sync happens on Theme Settings save. Did you pressed the Save button?

**Comment by danyj** [comment]
is not even seeing the change , or me changing the input 


here , 


this is complete customizer 


	<?php
	if (!defined('FW'))
		die('Forbidden');
	$options = array(
		'panel_id' => array(
			'type' => 'tab',
			'title' => 'Site',
			'options' => array(
				// This MUST contain 'box' containers
				'section_id' => array(
					'type' => 'box',
					'title' => 'Layout',
					'options' => array(
						// This MUST contain options (not containers)
						'site_width' => array(
							'type' => 'short-text',
							'value' => '1400px',
							'label' => __('Site max width', 'fw'),
							'desc' => __('Define website max width', 'fw'),
							'help' => __('You can use pixels ( px ) or percentage ( % ). If no operator is added px is default.', 'fw')
						),
					)
				)
			)
		)
		
	);

**Comment by moldcraft** [comment]
I can't reproduce, [it works](https://vid.me/TJah) on my side.

**Comment by danyj** [comment]
ok I got it , kirki plugin was on and had js error , disabled , 

on first load the input is empty 
if changed , change trigger via refresh but not visible 
admin option is change on save , so that is ok , 

but this is bugging me ,  if the refresh is initiated , why is it not changing my preview 


this option is outputted in the 

    wp_add_inline_style


via 

    fw_get_db_settings_option('site_width')

**Comment by moldcraft** [comment]
> but this is bugging me , if the refresh is initiated , why is it not changing my preview

The preview is not save, save happens only when the top Save button is pressed.
The preview changes only customizer values.
Try to replace `fw_get_db_settings_option('site_width')` with `fw_get_db_customizer_option('site_width')`

**Comment by danyj** [comment]
nice one!


		global $wp_customize;
		if (isset($wp_customize) && $wp_customize->is_preview() && ! is_admin() ) {
			
			
			$customizer_option = fw_get_db_customizer_option($id);
			if(isset($customizer_option)){
			
				return fw_get_db_customizer_option($id,$default);
			
			}else{
				
				fw_get_db_settings_option($id,$default);
				
			}
		}


fixes that for me

**Comment by thednp** [comment]
I was expecting this to work different. 

I haven't tested but according to what I read here, was thinking that customizer only updates the theme options values and use them (or the essential part of them) and not work separate. Handling 2 option getters <code>fw_get_db_customizer_option()</code> and <code>fw_get_db_settings_option()</code>. 

Am I wrong or ask too much of it?

**Comment by danyj** [comment]
I asked that to 4 days ago , 
https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-100484833

but @moldcraft  wants to keep default wp functionality for devs who just want to use get_mods option 
and not using theme admin at all.

**Comment by danyj** [comment]
In my case I was lucky that when I started the work 5 months ago I made my own get option function 

    thz_get_option(
      // some conditions


so I was able to switch very fast to customizer options , otherwise it would have been a workaround

**Comment by thednp** [comment]
I have same feature, but I think I need to change it to handle these huge arrays

<pre>
function get_my_option(){
   //now I need to search the array for the option key, before I didn't need it, the framework was different
}
</pre>

**Comment by danyj** [comment]
this worked fine for me 

https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-101642019

but I used the default way fw does things without data changes

if I had  big array coming from 


    fw_get_db_settings_option($id,$default);


same array in preview comes from 


    fw_get_db_customizer_option($id,$default);

**Comment by danyj** [comment]
again , not 100% tested , just few basic options for now, I hope il get to full customizer in 2 weeks.

**Comment by thednp** [comment]
So for the sake of the composer, I need to redo most of my 3500+ line again, in another file, and build it's JS also right?

**Comment by danyj** [comment]
@thednp  kinda yes , this is why I said this 
https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-100935709

**Comment by moldcraft** [comment]
The design is ok.
So there is nothing else to do at this task.
I will create a new Unyson release for approval.

**Comment by thednp** [comment]
Quick questions: 

1. does the framework do any JS scripting for the customizer?
2. if using LESS compile at theme options Save, how the customizer will deal with it?

**Comment by moldcraft** [comment]
1. Frontend - no. Backend - [yes](https://github.com/ThemeFuse/Unyson/blob/master/framework/static/js/backend-customizer.js).
2. Customizer preview affects/replaces only the `fw_get_db_customizer_option()` result. So in preview  `fw_get_db_settings_option()` will return "old" db value. If you are using [this script](https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-101633999) only after customizer Save `fw_get_db_settings_option()` will be affected.

I undestand that the problem with Customizer<->ThemeSettings sync is only for themes that were already released and want to move the Theme Settings options to Customizer (due to TF requirement). But if you are still building your theme (it is not released yet) why you are using both Theme Settings and Customizer, instead of moving everything to Customizer?

**Comment by moldcraft** [comment]
> if using LESS compile at theme options Save, how the customizer will deal with it?

Better answer: If you are using [this action](https://github.com/ThemeFuse/Unyson/blob/6ae3909793499c4ff88c5c85ceaba8fb5059ad92/framework/core/components/backend.php#L1137) to compile the LESS, on customizer Save it is not triggered. So if you are using [this script](https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-101633999), the db settings options will be updated but the LESS will not be recompiled. You can execute the LESS compile at the end of `_action_after_customizer_save()` method (in that script).

**Comment by thednp** [comment]
What about changing theme layout from customizer? Perhaps this wouldn't be possible, would it force the theme to load different PHP files? I guess not.

**Comment by moldcraft** [comment]
Do you mean frontend theme to be different from customizer preview theme?
*I feel it's not a god idea to do this kind of things.*

**Comment by thednp** [comment]
So customizer basically works for most preview related stuff like: colors, logo (image/s and spacing), and NOT for theme layout (when using different layouts with different php files), typography, etc. 

Basic stuff. 

And this would only work as INLINE style for preview, the LESS compile would happen when customizer save is triggered, as you described.

**Comment by moldcraft** [comment]
> and NOT for theme layout (when using different layouts with different php files), typography

It should work for this cases, because the Theme Preview is an iframe and it makes a full page refresh.

> the LESS compile would happen when customizer save is triggered, as you described.

yes

**Comment by moldcraft** [comment]
https://github.com/ThemeFuse/Unyson-Extensions-Approval/issues/66

http://manual.unyson.io/en/latest/options/introduction.html#customizer 
*P.S. If your English is better, you can always contribute with a [pull request](https://github.com/ThemeFuse/Unyson-Documentation/pulls)*

**Comment by duongancol** [comment]
@moldcraft : i got this issue, typography option did not return the value default in customizer. 
When I use this code:
<pre>
'demo_typography'                => array(
                        'label' => __( 'Typography', 'unyson' ),
                        'type'  => 'typography',
                        'value' => array(
                            'size'   => 17,
                            'family' => 'Verdana',
                            'style'  => '300italic',
                            'color'  => '#0000ff'
                        ),
                        'desc'  => __( 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
                            'unyson' ),
                        'help'  => sprintf( "%s \n\n'\"<br/><br/>\n\n <b>%s</b>",
                            __( 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
                                'unyson' ),
                            __( 'Sed ut perspiciatis, unde omnis iste natus error sit voluptatem accusantium doloremque laudantium',
                                'unyson' )
                        ),
                    ),
</pre>
It showed http://i.imgur.com/NAS97ni.png

**Comment by duongancol** [comment]
It's also with pick color option, please check it!

**Comment by duongancol** [comment]
I also suggest a Reset button in Customizer. It's the same function with Reset button in Theme Setting. 
It's necessary too!

**Comment by moldcraft** [comment]
I made [the fix](https://github.com/ThemeFuse/Unyson/commit/b42f2933df1c830826d67ea92f0e9410195b989c). Test it please.

**Comment by duongancol** [comment]
The typography and color option also can get default value now. I think it's fixed. 

How about a reset button?

**Comment by giathinh910** [comment]
I have a question about get the default value.

```php
'demo_multi_picker_radio'        => array(
	'type'         => 'multi-picker',
	'label'        => false,
	'desc'         => false,
	'value'        => array(
		'gadget' => 'laptop',
	),
	'picker'       => array(
		'gadget' => array(
			'label'   => __( 'Multi Picker: Radio', 'unyson' ),
			'type'    => 'radio',
			'choices' => array(
				'phone'  => __( 'Phone', 'unyson' ),
				'laptop' => __( 'Laptop', 'unyson' )
			),
		)
	),
	'choices'      => array(
		'phone'  => array(
			'price'  => array(
				'type'  => 'text',
				'label' => __( 'Price', 'unyson' ),
                                'value' => '10'
			),
		),
		'laptop' => array(
			'price'  => array(
				'type'  => 'text',
				'label' => __( 'Price', 'unyson' ),
                                'value' => '100'
			),
		),
	),
),
```

With these config, I can get default value of [demo_multi_picker_radio][gadget] = 'laptop'. But laptop's price is still null. Is there any way to make the default value of choices appear. Thank you.

**Comment by moldcraft** [comment]
> @duongancol How about a reset button?

This is not our custom made Theme Settings page where we can do anything we want, the Customizer page is created by WordPress and we can use only filters and actions to do what we want.

In customizer are registered many [settings](https://codex.wordpress.org/Theme_Customization_API#Adding_a_New_Setting), each has a [`type` argument](https://codex.wordpress.org/Class_Reference%5CWP_Customize_Manager%5Cadd_setting#Arguments) which indicates where that setting is saved. `type` can be a custom value that can be handled using hooks: [1](https://github.com/WordPress/WordPress/blob/33b61a9244a6b9dd7d64477483f729b1ea607984/wp-includes/class-wp-customize-setting.php#L399), [2](https://github.com/WordPress/WordPress/blob/33b61a9244a6b9dd7d64477483f729b1ea607984/wp-includes/class-wp-customize-setting.php#L324). In order to Reset/Delete all setting values, I should get/collect them all and somehow call a `delete/reset()` method.

I think there is an action on customizer html render, and I can use it to output the Reset button. 
But it's not clear how to replace controls html on Reset click, also how to get that html.
In Theme Settings, on Reset, we do an ajax that deletes all settings from db, then another ajax (GET request) to the same page to generate on server side the form's html with new values, then use it to replace current form html. Then we do `fwEvents.trigger('fw:options:init', {$elements: $('<new form html>')}` to init options.

https://github.com/ThemeFuse/Unyson/issues/576

**Comment by danyj** [comment]
@moldcraft  @duongancol  how about  a reset button option rather than default button show?
I might not want to have the reset button but could use it in some other theme. an option would make it easier to collect things you need and will give us additional flexibility

**Comment by duongancol** [comment]
@moldcraft I hope you can do that
@danyj : I'm not sure, it's ok if your theme has some options in customizer, but with my theme, more than 20+ option that include typography, color, multi-select. It's really necessary features.

**Comment by danyj** [comment]
@duongancol , I lost you there bud , did not understand quite what you mean. 
but what I mean is a reset button as and option type that you can just place in customizer

**Comment by duongancol** [comment]
@danyj I mean i'm not sure about others, but I think it's necessary for my theme.

**Comment by moldcraft** [comment]
@giathinh910 I made [the fix](https://github.com/ThemeFuse/Unyson/commit/96e2ab2a02e71e205f04838e3a9ece6c239c4fd9). Please test it.

**Comment by giathinh910** [comment]
I've just test customizer on a fresh installation. The bug that I I can't get the default value inside multi-picker option is fixed. The initial option working perfectly. However the issue that I update in the previous still exist that when I change one field in any section and make save, all other option become null in front page. Hope you can check it soon. Thank you very much.

**Comment by moldcraft** [comment]
> However the issue that I update in the previous still exist that when I change one field in any section and make save, all other option become null in front page. 

Can you make a video or give me the steps to reproduce this?

**Comment by giathinh910** [comment]
Here is my steps.
Initially, after a fresh installation. All the value can be get perfectly without touching any where. You can notice a value that I take for example.

![1_caption](https://cloud.githubusercontent.com/assets/6775636/7746731/004e7bde-ffdf-11e4-8b97-f115b42e0366.gif)

Then I open customizer, change that value. From the customizer's view, you can still see that value is changed and others stay till.

![2_caption](https://cloud.githubusercontent.com/assets/6775636/7746755/330bab6e-ffdf-11e4-84f6-0c94f4ee0cbc.gif)

However when going back to front page, you could see that the example value changed but all other value gone, and I get tons of undefined variable notice.

![3_caption](https://cloud.githubusercontent.com/assets/6775636/7746789/6daf2bc4-ffdf-11e4-9124-d9e27519904f.gif)

**Comment by moldcraft** [comment]
@giathinh910 I made [another fix](https://github.com/ThemeFuse/Unyson/commit/87464e59dee6e9d2cc9a10280a35f4839c9b409f). Please test.

**Comment by giathinh910** [comment]
It works nicely by now. Thank you very much :+1:

**Comment by moldcraft** [comment]
Thank you for reporting this.

**Comment by thednp** [comment]
I'm still puzzled on how to actually do LESS on each change, I mean I know I can hook the LESS compile function to compozer change event, but If user cancels, how would revert back to saved settings and recompile the CSS again?

That's a pretty tough issue, I believe I cannot do composer after all.

**Comment by moldcraft** [comment]
@thednp Sorry, I don't know a solution for you problem with LESS. Doing a recombile on every preview refresh seems not a good idea.

**Comment by duongancol** [comment]
when I set color picker option in the bottom of section, it'll be hidden if I have another section. Please see my attachment image for detail: http://prntscr.com/7cwds4

**Comment by giathinh910** [comment]
Hi. Is there anyway to hide the font size field in the typoghraphy option because I always let user change font size base on section, not base on font-family. Thank you.

![screenshot_1](https://cloud.githubusercontent.com/assets/6775636/8001265/36dc1548-0b8e-11e5-90bd-f27918773d07.jpg)

**Comment by moldcraft** [comment]
https://github.com/ThemeFuse/Unyson/issues/640

**Comment by duongancol** [comment]
hi @moldcraft 
is there any ways to add Icon to Customizer ? 
For example here: http://prntscr.com/7t8ciq

**Comment by moldcraft** [comment]
Unyson has nothing to do with Customizer sections, ask WP developers.

**Comment by albertso** [comment]
I'm trying to use Styling-Extension options in the customizer.

The values of 'style' options are being stored as expected in fw_options, but CSS styles are not generated/stored in extension data.

CSS generation is done with FW_Styling_Css_Generator->[get_css()](https://github.com/ThemeFuse/Unyson-Styling-Extension/blob/master/includes/class-fw-styling-css-generator.php#L21), so I **could** use that to generate css on every page render, but the desired behaviour would be to let the extension generate/store/display the css for _all_ 'style' options, wherever they are (like in the customizer), and use [fw_ext_styling_get()](http://manual.unyson.io/en/latest/extension/styling/index.html#helpers) to get the data.

Currently CSS generation happens only for 'style' options that are in the Appearance->Styling page, declared in `extensions/styling/options/appearance-settings.php`, as you can see in FW_Extension_Styling->[_admin_action_generate_css()](https://github.com/ThemeFuse/Unyson-Styling-Extension/blob/master/class-fw-extension-styling.php#L162)

	public function _admin_action_generate_css() {
		$theme_options         = fw_extract_only_options( $this->get_options('appearance-settings') );
		$saved_data            = fw_get_db_extension_data( $this->get_name(), 'options' );
		$css_for_style_options = FW_Styling_Css_Generator::get_css( $theme_options, $saved_data );
		fw_set_db_extension_data( $this->get_name(), 'css', $css_for_style_options );
	}

This action is hooked to 'fw_styling_form_save', in FW_Extension_Styling->[add_admin_actions()](https://github.com/ThemeFuse/Unyson-Styling-Extension/blob/master/class-fw-extension-styling.php#L44):

		add_action( 'fw_' . $this->get_name() . '_form_save', array( $this, '_admin_action_generate_css' ) );

I haven't found where it should be hooked to affect customizer options, but I guess something like 'fw_option_type_style_save'... ?

**Comment by llemurya** [comment]
@albertso we make options to work in customizer is the same as styling and the styling will be deprecated.Please read our documentation about [options in customizer](http://manual.unyson.io/en/latest/options/introduction.html#customizer).It is not correct to use styling option type in customizer because the styling is more then an option.

**Comment by moldcraft** [comment]
> https://github.com/ThemeFuse/Unyson/issues/410#issuecomment-108703971

We inspected the customizer html and didn't find a solution to make color picker go above (z-index doesn't help). If you know a solution, please share.

----

P.S. [ReTweet](https://twitter.com/ThemeFuse/status/624513526128885760) if you like Unyson Options in Customizer.

**Comment by danyj** [comment]
@moldcraft  
http://prntscr.com/7ylgbe
switch it to relative and it will push down

**Comment by danyj** [comment]
@moldcraft  here video  ,  http://screencast.com/t/DGj8Fhmh0  
you can do that only if in customizer

**Comment by duongancol** [comment]
I think the best solution is reducing margin-top of picker. It'll look like that: http://prntscr.com/7yphlu

**Comment by moldcraft** [comment]
@danyj @duongancol Thank you

**Comment by moldcraft** [comment]
[It's already fixed](https://vid.me/KwWP)

**Comment by huykon** [comment]
How can I save Customizer with LESS and then compile to CSS? Anyone help me.

**Comment by RussellAzim** [comment]
How i integrate unyson customizer
